src/libstore/derivations.cc

```diff
@@ -103,31 +103,66 @@ StorePath writeDerivation(Store & store,
 }
 
 
-/* Read string `s' from stream `str'. */
-static void expect(std::istream & str, std::string_view s)
-{
-    char s2[s.size()];
-    str.read(s2, s.size());
-    if (std::string(s2, s.size()) != s)
```

Comment:
Just make this `std::string_view`. After that, we can see if any further benchmarking is worth it, but I doubt it.

Reply:
Can you clarify what I should change `str.read()` to? Note that `str` is the `std::istream`.

Reply:
Also note that initial profile shows `istream::get()` as a big part of it. That probably means it's heavyweight implementation that includes exceptions handling does not inline well.

Reply:
we've also seen that while optimizing the evaluator. istreams are *incredibly* expensive if what you really want is a char iterator. just getting rid of that saves a whole bunch of non-inlineable virtual calls and other overhead, that's definitely worth doing.

Reply:
The `istringstream_nocopy` part solved that.